daft/io/_lance.py

```diff
@@ -0,0 +1,120 @@
+# isort: dont-add-import: from __future__ import annotations
+
+from typing import TYPE_CHECKING, Iterator, List, Optional
+
+from daft import context
+from daft.api_annotations import PublicAPI
+from daft.daft import IOConfig, Pushdowns, PyTable, ScanOperatorHandle, ScanTask
+from daft.dataframe import DataFrame
+from daft.io.object_store_options import io_config_to_storage_options
+from daft.io.scan import PartitionField, ScanOperator
+from daft.logical.builder import LogicalPlanBuilder
+from daft.logical.schema import Schema
+from daft.table import Table
+
+if TYPE_CHECKING:
+    import lance
+
+
+def _lancedb_table_factory_function(
+    fragment: "lance.LanceFragment", required_columns: Optional[List[str]]
+) -> List["PyTable"]:
```

Comment:
instead of a `list[PyTable]`, you should have this return an `Iterator[PyTable]`. This will allow you to early terminate based off some control flow in rust.